---
layout: post
title: PlayData Day 8
description: >
    
# image: /assets/img/pythonLogo.png
sitemap: false
categories:
  - studylog
---

# PlayData AI Bootcamp Day 8

* this unordered seed list will be replaced by the toc
{:toc}

## filter
- iterable의 각 요소에 대해 function이 참인 것만 반환하는 요소의 이터레이터
  
```python
filter(function, iterable)
```
```python
ex = [1, 2, 3, 4, 5]
list(filter(lambda x: x%2 == 0, ex))
>>> [2,4]
```

## Enumerate & Zip
- **Enumerate**
	- 리스트의 요소를 추출할 때 번호를 붙여서 추출
	```python
	enumerate(변수)
	```

```python
for i, v in enumerate(['사과', '바나나', '딸기']):
	print(i)
	
>>> 0 사과
>>> 1 바나나
>>> 2 딸기
```

- `enumerate`를 리스트와 같이 쓰면 번호 순서와 각 리스트 원소를 tuple로 묶어서 돌려준다

```python
list(enumerate(['사과', '바나나', '딸기']))
>>> [(0, '사과'), (1, '바나나'), (2, '딸기')]
```

- `dictonary`로도 변환 가능

```python
dict(enumerate['사과', '바나나', '딸기'])

```


- **Zip**
	- 두개 이상의 리스트를 병렬적으로 추출
	```python
	zip(list1, list2, ...)
	```

```python
alist = ['사과', '바나나', '딸기']
blist = ['Apple', 'Banana', 'Strawberry']

dict(zip(alist, blist))

>>> {'딸기': 'Strawberry', '바나나': 'Banana', '사과': 'Apple'}
```

- 두개 이상의 리스트를 묶을 떄

```python
# zip
list(zip((1,2,3), (10,20,30), (100,200,300)))
>>> [(1, 10, 100), (2, 20, 200), (3, 30, 300)]
```

## reduce
- `reduce`는 반복가능한 객체의 각 요소를 지정된 함수로 처리한 뒤 이전 결과와 **누적해서 반환** 
- `reduce`는 Python3 부터 내장함수가 아닙니다
	- `functools`모듈에서 `reduce` 함수를 가져와야 합니다

```python
from functools import reduce
reduce(funciton, iterable, initializer=None)
```

```python
from functools import reduce
array2dim = [[x]*x for x in range(5)]

array = reduce(lambda x, y: x+y, array2dim)
array
>>> [1, 2, 2, 3, 3, 3, 4, 4, 4, 4]

```

```python
test2 = [{'name' : 'a', 'age' : 1}, {'name' : 'b', 'age' : 2}, {'name' : 'c', 'age' : 3}]

reduce((lambda x,y : x+y['age']), test2, 0) #초가값을 정해주어야 함
```
{% include pythontutor_reduce.md %}{:.centered }

## 모듈 (Module)
- 이미 만들어진 Python 소스파일(라이브러리)
	- 파이썬은 모듈을 만들기 위한 추가 과정은 필요하지 않습니다.

- 모듈은 어디서 가져올까요?
  - 모듈은 파이썬 소스코드에서 가져옵니다
  - 모듈은 위치를 선정해줘야 됩니다
  - 모듈은 위치를 미리 지정해놓습니다
  - 모듈은 항상 그 위치에서 가져오도록 만들어놓습니다


### 커스텀 모듈
- 위치를 확인하기 위해 다른 모듈을 사용
  ```python
  import sys
  sys.path
  ```


- `sys` 모듈
	- 파이썬에서 시스템 관리 관련 함수를 모아둔 모듈
	- `sys.path` : 모듈의 검색경로 리스트를 불러옴
	```python
	import sys
	sys.path
	
	>>> ['',
		 '/content',
		 '/env/python',
		 '/usr/lib/python37.zip',
		 '/usr/lib/python3.7',
		 '/usr/lib/python3.7/lib-dynload',
		 '/usr/local/lib/python3.7/dist-packages',
		 '/usr/lib/python3/dist-packages',
		 '/usr/local/lib/python3.7/dist-packages/IPython/extensions',
		 '/root/.ipython']
	```


- 모듈 가져오기

```python
import modules.myarea as mma
```

  - from을 이용해서 모듈에서 쓰려고 하는 내용을 직접 가져오는 법

    ```python
    from modules import func
    ```

  - 모듈에 있는 모든 기능을 가져오기 (소스코드 참조하기)
    
    ```python
    from modules import *
    ```

### Python 소스파일의 실행
- Python에서는 모듈, 일반 실행파일 모두 `.py`로 저장됩니다
- Python 소스코드를 직접 실행할 떄와 참조해서 실행할 댸와 구분하는 방법이 있습니다
  - 자신이 직접 실행할 때
    ```python
    __name__ == '__main__'
    ```
  - 모듈로 쓰여서 실행할때
    ```python
    __name__ == 모듈명  
    ```
### 패키지 (Package)
- 모듈들을 하나로 묶어서 관리하는 것
  - 파이썬 패키지는 폴더로 관리됩니다
  - 패키지의 이름은 폴더의 이름이 됩니다

## 예외처리 (Exception)
- 프로그램 실행 중에 예외가 발생하면 보통 '비정상종료'가 됩니다.
- 예외처리는 비정상 종료를 막기 위해 에러가 빌생된 이후의 문장은 정상적으로 수행하도록 처리하는 방방법입니다
- 파이썬 에러의 종류들
	![PythonErrors](/assets/img/PlayDataNote/Day8/pythonError.png){:.centered width="400" loading="lazy"}
```python
try:
	예외가 발생할 가능성이 있는 코드
except:
	예외가 발생할 떄 실행할 코드
except:
	예외가 발생할 떄 실행할 코드
else:
	잘 실행되면 실행할 코드
finally: 
	되든 안되든 반드시 실행할 코드
```

### 예외처리 방법 2가지
1. LBYL (Look Before You Leap) - 에러가 나기 전에 처리하기
	- 코드 실행 전에 예외 발생 없게 미리 검증하고 실행하기
	- 현실적으로 완벽히 에러를 예측하기 어려움
2.  EAFP (It's Easier to Ask Forgiveness than Permission)
	- 일단 코드를 실행하고 예외가 발생하면 그때 처리하기
	- Python은 EAFP 예외처리 권장
{:note}
- **TDD** 확인하기

#### LBYL 방식의 예외 처리
```python
def convert(string):
	return int(string)

convert('a')
```

```
>>> ValueError: invalid literal for int() with base 10: 'a'
```
- LBYL 예외 처리 사용 경우
```python
def convert(string):
	if isinstance(string, str):
		if string[0] == '-' or string[0] == '+' or string[0].isdecimal():
			if string[1:].isdecima(): return int(string)
			else: return False
		else: return False
	else: return False

convert('a')
>>> False
```

#### EAFP 방식의 예외 처리
```python
def convert(string):
	try:
		return(int(string))
	except:
		return False
```

- 어느 한 가지 방법만 이용하지는 않습니다
	- 두 방식 (LBYL, EAFP)들을 적절히 활용하는 것이 바람직합니다

### 다양한 error 처리 
```python
def convert(string):
	try:
		return(int(string))
	except ValueError as e:
		return ('에러 발생', e)
	except TypeError as e:
		return ('에러 발생', e)
```

```
convert('a')
>>> ('에러 발생', ValueError("invalid literal for int() with base 10: 'a'"))
```

```
convert([])
>>> ('에러 발생',
 TypeError("int() argument must be a string, a bytes-like object or a number, not 'list'"))
```


### 에러 객체 구조

#### else
- 잘 처리된 경우 실행

#### finally
- 뭐가 됬든 실행

```python
def convert(string):
  try:
    print( int(string) )
  except ValueError as e:
    print('ValueError', e)
  except Exception as e:
    print('Unkown Exception', e)
  else:
    print('예외가 발생하지 않으면 실행')
  finally:
    print('예외 유무와 상관없이 마지막에 항상 무조건 실행')
```

```
convert('a')
>>> ValueError invalid literal for int() with base 10: 'a'
	예외 유무와 상관없이 마지막에 항상 무조건 실행
```

```
convert('123')
123
>>> 예외가 발생하지 않으면 실행
	예외 유무와 상관없이 마지막에 항상 무조건 실행
```

### 예외처리 응용
#### EOF (End of File)
- 파일의 가장 마지막에 들어가는 문자
- 파이썬은 EOF처리를 잘 못함
	- 언어의 끝을 예외로 처리 (EOFError)
	- 종료를 의미하는 단축키
		- `ctrl + z`
		
```python
while True:
	read = input()
	print(read)
	
```
#### EOL(End of Line)

### 처리되지 않는 예외

- Compile Error: 
	- 파이썬같은 인터프리터 언어에서 [REPL](/_posts/studylog/2022-03-14-1st-day.md##REPL): Read -> Evaluaiton 사이(번역) 할 때 발생하는 에러
		- 문법적인 오류는 이 과정에서 발생하기 때문에 예외처리가 되지 않습니다
- Runtime Error: 프로그램 실행 중인 상태에서 발생하는 에러
	- 번역은 끝난 상태에서 실행하는 더중에 문제가 발생하는 경우
	- Evaluaiton 중에 발생

### 임의 에러
- 사용자가 임의로 에러를 만들어서 에러의 클래스까지 따로 만들어서 처리합니다
- 새로 에러를 만드로 싶을 때는 해당 클래스를 만들어줘야 합니다.
	- 각각의 에러틑 하나의 클래스(묶음)으로 만들어져 있고, 그 묶음은 Exception이라는 상위 클래스를 상속받는다.
	-   Exception이라는 클래스를 상속받는 새 에러 묶음을 만들어서 처리한다.
#### raise

```python
class NegativeDivisionError(Exception):
	def __init__(self, value):
		self.value = value


def PositiveDivide(a, b):
	if (b<0):
		raise NegativeDivisionError(b) # 사용자 임의 에러
	return a/b

try:
	ex = PositiveDivide(3, -3)
except ZeroDivisionError as e:
	print(e)
except NegativeDivisionError as e:
	print(e)
except Exception as e:
	print(e)


PositiveDivide(2, -1)
```

```
>>> NegativeDivisionError: -1
```
